src/java.desktop/windows/native/libawt/windows/ShellFolder2.cpp

```diff
@@ -1057,44 +1057,57 @@ JNIEXPORT jintArray JNICALL Java_sun_awt_shell_Win32ShellFolder2_getIconBits
-            if (IS_WINXP) {
-                for (int i = 0; i < nBits; i++) {
-                    if ((colorBits[i] & 0xff000000) != 0) {
-                        hasAlpha = TRUE;
-                        break;
+            long *colorBits = NULL;
+            try {
+                entry_point();
+                colorBits = (long*)safe_Malloc(MAX_ICON_SIZE * MAX_ICON_SIZE * sizeof(long));
+            } catch(std::bad_alloc&) {
+                handle_bad_alloc();
+            }
+
+            if (colorBits != NULL) {
+                GetDIBits(dc, iconInfo.hbmColor, 0, iconSize, colorBits, &bmi, DIB_RGB_COLORS);
+                // XP supports alpha in some icons, and depending on device.
+                // This should take precedence over the icon mask bits.
+                BOOL hasAlpha = FALSE;
+                if (IS_WINXP) {
+                    for (int i = 0; i < nBits; i++) {
+                        if ((colorBits[i] & 0xff000000) != 0) {
+                            hasAlpha = TRUE;
+                            break;
+                        }
                     }
                 }
-            }
-            if (!hasAlpha) {
-                // Extract the mask bitmap
-                long *maskBits = (long*)safe_Malloc(MAX_ICON_SIZE * MAX_ICON_SIZE * sizeof(long));
-                GetDIBits(dc, iconInfo.hbmMask, 0, iconSize, maskBits, &bmi, DIB_RGB_COLORS);
-                // Copy the mask alphas into the color bits
-                for (int i = 0; i < nBits; i++) {
-                    if (maskBits[i] == 0) {
-                        colorBits[i] |= 0xff000000;
+                if (!hasAlpha) {
+                    // Extract the mask bitmap
+                    long *maskBits = NULL;
+                    try {
+                        entry_point();
+                        maskBits = (long*)safe_Malloc(MAX_ICON_SIZE * MAX_ICON_SIZE * sizeof(long));
+                    } catch(std::bad_alloc&) {
+                        handle_bad_alloc();
+                    }
+                    if (maskBits != NULL) {
+                        GetDIBits(dc, iconInfo.hbmMask, 0, iconSize, maskBits, &bmi, DIB_RGB_COLORS);
+                        // Copy the mask alphas into the color bits
+                        for (int i = 0; i < nBits; i++) {
+                            if (maskBits[i] == 0) {
+                                colorBits[i] |= 0xff000000;
+                            }
+                        }
+                        free(maskBits);
                     }
                 }
-                free(maskBits);
-            }
-            // Release DC
-            ReleaseDC(NULL, dc);
-            // Create java array
-            iconBits = env->NewIntArray(nBits);
-            if (!(env->ExceptionCheck())) {
-                // Copy values to java array
-                env->SetIntArrayRegion(iconBits, 0, nBits, colorBits);
+                // Release DC
+                ReleaseDC(NULL, dc);
```

Comment:
The DC has to be released even if `colorBits` allocation failed, so this needs to be after `if (colorBits != NULL)`.

Reply:
Fixed.

Reply:
Probably, you hasn't pushed the change for releasing the DC.

Reply:
Indeed.